### PR TITLE
Improve thread safety in PushManager and NotificationsManager.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/environment/NotificationsManager.m [hunk 1]
```diff
@@ -211,31 +211,33 @@ - (NotificationType)notificationPreviewType
...
-
-    // Replace any existing notification
-    // e.g. when an "Incoming Call" notification gets replaced with a "Missed Call" notification.
-    if (self.currentNotifications[identifier]) {
-        [self cancelNotificationWithIdentifier:identifier];
-    }
+    dispatch_async(dispatch_get_main_queue(), ^{
+        // Replace any existing notification
+        // e.g. when an "Incoming Call" notification gets replaced with a "Missed Call" notification.
+        if (self.currentNotifications[identifier]) {
```
Comment (charlesmchen) [inline Signal/src/environment/NotificationsManager.m]: Only modify `currentNotifications` on the main thread.

File: Signal/src/network/PushManager.m [hunk 2]
```diff
@@ -431,32 +431,34 @@ - (BOOL)applicationIsActive {
...
+            // but the more lag we introduce to notification delivery.
+            const CGFloat kDelaySeconds = 0.3f;
+            notification.fireDate = [NSDate dateWithTimeIntervalSinceNow:kDelaySeconds];
+            notification.timeZone = [NSTimeZone localTimeZone];
+        }
 
-    [[UIApplication sharedApplication] scheduleLocalNotification:notification];
-    [self.currentNotifications addObject:notification];
+        [[UIApplication sharedApplication] scheduleLocalNotification:notification];
+        [self.currentNotifications addObject:notification];
```
Comment (charlesmchen) [inline Signal/src/network/PushManager.m]: Only modify `currentNotifications` on the main thread.

**Comment by charlesmchen** [comment]
I guess I'm also wondering if there's some redundancy between these two managers that could be de-duplicated.

**Comment by charlesmchen** [comment]
> Is the point of this change that though we're already assert'ing MainThread, that we're not running this code in DEBUG often, so you don't have confidence in their assertion?

These two managers mutate their `currentNotifications` collections in these methods, which isn't safe to do concurrently.  `[PushManager cancelNotificationsWithThreadId:]` in particular iterates over this collection then mutates it which is not safe if another thread is mutating the collection at the same time.

**Comment by michaelkirk** [comment]
But if we're already asserting that we're on the main thread whenever we read or edit, how could it have ever happened concurrently?

**Comment by charlesmchen** [comment]
> But if we're already asserting that we're on the main thread whenever we read or edit, how could it have ever happened concurrently?

I just added those asserts last week, wondering if this is an issue.  Then I hit the asserts, hence this PR.  Gotta love asserts!

**Comment by michaelkirk** [comment]
Well that clears it up. LGTM

**Pull request merged**